### PR TITLE
feat: Table navigation using TAB key

### PR DESCRIPTION
**Pull request by AnsahMohammad**
Implemented the navigation of Table cells using TAB key

This PR solves [[FR] Table Cells navigation using TAB key #3982](https://github.com/AppFlowy-IO/AppFlowy/issues/3982)

### CODE REVIEW

**Comment by Xazin** [comment]
Would be nice to also have Shift+tab to move the reverse way.

**Comment by AnsahMohammad** [comment]
> Would be nice to also have Shift+tab to move the reverse way.

working on that now,
image

Shift+tab isn't working for some reason,

**Comment by Xazin** [comment]
> > Would be nice to also have Shift+tab to move the reverse way.
> 
> working on that now, image
> 
> Shift+tab isn't working for some reason,

Maybe `shift+tab` in lowercase?

**Comment by AnsahMohammad** [comment]
> Maybe `shift+tab` in lowercase?

yeah, the issue was, that `hot reload` wasn't working so I was confused. It is done now :+1:

**Comment by AnsahMohammad** [comment]
Hi @Xazin, could you review my PR when you get a chance? Thank you!

**Comment by LucasXu0** [comment]
@AnsahMohammad Can you add the tests to cover the code?

**Comment by AnsahMohammad** [comment]
@LucasXu0 I have added tests for Tab, Shift+Tab, Arrow Right and Arrow Left.

**Comment by AnsahMohammad** [comment]
> @AnsahMohammad Can you add the tests to cover the code?

ready for review :+1:

**Comment by LucasXu0** [comment]
> > @AnsahMohammad Can you add the tests to cover the code?
> 
> ready for review 👍

Nice! I will check it now.

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 1]
```diff
@@ -152,6 +166,42 @@ CommandShortcutEventHandler _downInTableCellHandler = (editorState) {
   return KeyEventResult.ignored;
 };
 
+CommandShortcutEventHandler _tabInTableCellHandler = (editorState) {
+  final inTableNodes = _inTableNodes(editorState);
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:170]: Does the `_hasSelectionAndTableCell` always return true if the result of `_inTableNodes` is not empty?

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 2]
```diff
@@ -152,6 +166,42 @@ CommandShortcutEventHandler _downInTableCellHandler = (editorState) {
...
+  if (_hasSelectionAndTableCell(inTableNodes, selection)) {
+    final nextNode = _getNextNode(inTableNodes, 1, 0);
+    if (_nodeHasTextChild(nextNode)) {
+      editorState.selectionService.updateSelection(
+        Selection.single(
+          path: nextNode!.childAtIndexOrNull(0)!.path,
+          startOffset: 0,
+        ),
+      );
+    }
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:181]:
```suggestion
    if (nextNode != null && _nodeHasTextChild(nextNode)) {
      final firstChild = nextNode.childAtIndexOrNull(0);
      if (firstChild != null) {
        editorState.selection = Selection.single(
          path: firstChild.path,
          startOffset: 0,
        );
      }
    }
```

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 3]
```diff
@@ -152,6 +166,42 @@ CommandShortcutEventHandler _downInTableCellHandler = (editorState) {
...
+  if (_hasSelectionAndTableCell(inTableNodes, selection)) {
+    final nextNode = _getPreviousNode(inTableNodes, 1, 0);
+    if (_nodeHasTextChild(nextNode)) {
+      editorState.selectionService.updateSelection(
+        Selection.single(
+          path: nextNode!.childAtIndexOrNull(0)!.path,
+          startOffset: 0,
+        ),
+      );
+    }
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:199]:
```suggestion
    final previousNode = _getPreviousNode(inTableNodes, 1, 0);
    if (previousNode != null && _nodeHasTextChild(previousNode)) {
      final firstChild = previousNode.childAtIndexOrNull(0);
      if (firstChild != null) {
        editorState.selection = Selection.single(
          path: firstChild.path,
          startOffset: 0,
        );
      }
    }
```

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 4]
```diff
@@ -191,15 +241,47 @@ bool _hasSelectionAndTableCell(
...
 
-Node? _getNextNode(Iterable<Node> nodes, int colDiff, rowDiff) {
+Node? _getNextNode(Iterable<Node> nodes, int colDiff, int rowDiff) {
   final cell = nodes.first.parent!;
   final col = cell.attributes[TableCellBlockKeys.colPosition];
   final row = cell.attributes[TableCellBlockKeys.rowPosition];
-  return cell.parent != null
-      ? getCellNode(cell.parent!, col + colDiff, row + rowDiff)
-      : null;
+  final table = cell.parent!;
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:248]: don't force unwrap the `parent`.

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 5]
```diff
@@ -191,15 +241,47 @@ bool _hasSelectionAndTableCell(
...
+
+  final numCols = table.children.last.attributes['colPosition'] + 1;
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:250]: use `TableCellBlockKeys.colPosition` instead.

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 6]
```diff
@@ -191,15 +241,47 @@ bool _hasSelectionAndTableCell(
...
+  final numRows = table.children.last.attributes['rowPosition'] + 1;
+
+  var nextCol = (col + colDiff) % numCols;
+  var nextRow = row + rowDiff + ((col + colDiff) ~/ numCols);
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:254]: Please add comments for them.

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 7]
```diff
@@ -191,15 +241,47 @@ bool _hasSelectionAndTableCell(
...
+  }
+}
+
+Node? _getPreviousNode(Iterable<Node> nodes, int colDiff, int rowDiff) {
+  final cell = nodes.first.parent!;
+  final col = cell.attributes[TableCellBlockKeys.colPosition];
+  final row = cell.attributes[TableCellBlockKeys.rowPosition];
+  final table = cell.parent!;
+
+  final numCols = table.children.last.attributes['colPosition'] + 1;
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:269]: same as previous one.

File: lib/src/editor/block_component/table_block_component/table_commands.dart [hunk 8]
```diff
@@ -191,15 +241,47 @@ bool _hasSelectionAndTableCell(
...
+  final numRows = table.children.last.attributes['rowPosition'] + 1;
+
+  var prevCol = (col - colDiff + numCols) % numCols;
+  var prevRow = row - rowDiff - ((col - colDiff) < 0 ? 1 : 0);
```
Comment (LucasXu0) [inline lib/src/editor/block_component/table_block_component/table_commands.dart:273]: Please add comments for them.

File: lib/src/editor/block_component/table_block_component/table_commands.dart (hunk 1 above)
Comment (AnsahMohammad) [reply lib/src/editor/block_component/table_block_component/table_commands.dart:170]: no, we check for the following conditions too :
```
   nodes.length == 1 &&
   selection != null &&
   selection.isCollapsed &&
   nodes.first.parent?.type == TableCellBlockKeys.type;
 ```

**Pull request merged**